File: WS_EUREKABANK_RESTFULLNOTNET_GR04/02.CLIESC/CLIESC_EurekaBank_REST_GR04/CLIESC_EurekaBank_REST_GR04/ec.edu.monster.services/RestEurekaApi.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using ec.edu.monster.models;

namespace ec.edu.monster.services
{
    /// <summary>
    /// Cliente REST contra el servidor ASP.NET (01.SERVIDOR).
    /// </summary>
    public class RestEurekaApi : IEurekaApi, IDisposable
    {
        private readonly HttpClient _http;

        private record MovimientoDto(string CuentaCodigo, int Numero, DateTime Fecha,
            string TipoCodigo, decimal Importe, string? ReferenciaCuenta);

        public RestEurekaApi(string baseUrl = null)
        {
            var url = string.IsNullOrWhiteSpace(baseUrl)
                ? "https://localhost:7043/"
                : baseUrl.Trim();
            _http = new HttpClient { BaseAddress = new Uri(url) };
        }

        public async Task<bool> ValidarIngreso(string usuario, string password)
        {
            var resp = await _http.PostAsJsonAsync("api/CoreBancario/validarIngreso",
                new { usuario, password });
            var texto = (await resp.Content.ReadAsStringAsync()).Trim();
            return texto.Equals("Exitoso", StringComparison.OrdinalIgnoreCase);
        }

        public async Task<List<Movimiento>> TraerMovimientos(string cuenta)
        {
            var dto = await _http.GetFromJsonAsync<List<MovimientoDto>>(
                $"api/CoreBancario/cuentas/{cuenta}/movimientos");

            return dto?
                .Select(m =>
                {
                    var (tipo, accion) = MapTipoAccion(m.TipoCodigo);
                    return new Movimiento
                    {
                        Cuenta = m.CuentaCodigo ?? string.Empty,
                        NroMov = m.Numero,
                        Fecha = m.Fecha,
                        Tipo = tipo,
                        Accion = accion,
                        Importe = m.Importe
                    };
                })
                .ToList() ?? new List<Movimiento>();
        }

        public async Task<OperacionCuentaResponse> RegDeposito(string cuenta, decimal importe)
        {
            return await PostOperacion("api/CoreBancario/deposito",
                new { cuenta, importe });
        }

        public async Task<OperacionCuentaResponse> RegRetiro(string cuenta, decimal importe)
        {
            return await PostOperacion("api/CoreBancario/retiro",
                new { cuenta, importe });
        }

        public async Task<OperacionCuentaResponse> RegTransferencia(string origen, string destino, decimal importe)
        {
            return await PostOperacion("api/CoreBancario/transferencia",
                new { cuentaOrigen = origen, cuentaDestino = destino, importe });
        }

        public Task<string> ProbarConexion() => Task.FromResult("OK");

        private async Task<OperacionCuentaResponse> PostOperacion(string path, object payload)
        {
            var resp = await _http.PostAsJsonAsync(path, payload);
            return await resp.Content.ReadFromJsonAsync<OperacionCuentaResponse>()
                   ?? new OperacionCuentaResponse { Estado = -1, Saldo = -1 };
        }

        private static (string Tipo, string Accion) MapTipoAccion(string codigo) =>
            codigo switch
            {
                "003" => ("DEPOSITO", "INGRESO"),
                "004" => ("RETIRO", "SALIDA"),
                "008" => ("TRANSFERENCIA", "INGRESO"),
                "009" => ("TRANSFERENCIA", "SALIDA"),
                "001" => ("APERTURA", "INGRESO"),
                _ => ("", "")
            };

        public void Dispose() => _http?.Dispose();
    }
}

File: WS_EUREKABANK_RESTFULLJAVA_GR04/02.CLIESC/CLIESC_EurekaBank_RESTJAVA_GR04/src/main/java/ec/edu/monster/service/RestEurekaApi.java
package ec.edu.monster.service;

import com.google.gson.*;
import ec.edu.monster.config.AppConfig;
import ec.edu.monster.model.Movimiento;
import ec.edu.monster.model.OperacionCuentaResponse;
import ec.edu.monster.util.Debug;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

/**
 * Cliente REST que consume el recurso CoreBancarioResource del servidor Java.
 */
public class RestEurekaApi implements EurekaApi {

  private final HttpClient http = HttpClient.newBuilder()
      .connectTimeout(Duration.ofSeconds(5))
      .build();

  private final Gson gson = new GsonBuilder()
      .setLenient()
      .create();

  private static String qs(Map<String, String> params) {
    if (params == null || params.isEmpty()) return "";
    StringJoiner sj = new StringJoiner("&");
    params.forEach((k, v) -> sj.add(URLEncoder.encode(k, StandardCharsets.UTF_8)
        + "=" + URLEncoder.encode(v, StandardCharsets.UTF_8)));
    return sj.toString();
  }

  private HttpResponse<String> send(HttpRequest req) throws IOException, InterruptedException {
    HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString());
    if (res.statusCode() >= 200 && res.statusCode() < 300) return res;
    throw new IOException(req.uri() + " -> " + res.statusCode() + " | " + res.body());
  }

  @Override
  public boolean validarIngreso(String usuario, String password) throws Exception {
    String uri = AppConfig.BASE_URL + "/login?" + qs(Map.of(
        "usuario", usuario,
        "password", password));

    Debug.log("POST " + uri);
    HttpRequest req = HttpRequest.newBuilder(URI.create(uri))
        .timeout(Duration.ofSeconds(10))
        .header("Accept", "application/json")
        .POST(HttpRequest.BodyPublishers.noBody())
        .build();
    JsonObject json = JsonParser.parseString(send(req).body()).getAsJsonObject();
    String res = json.has("resultado") ? json.get("resultado").getAsString() : "";
    Debug.log("Respuesta login=" + res);
    return "Exitoso".equalsIgnoreCase(res);
  }

  @Override
  public List<Movimiento> traerMovimientos(String cuenta) throws Exception {
    String uri = AppConfig.BASE_URL + "/movimientos/" +
        URLEncoder.encode(cuenta, StandardCharsets.UTF_8);

    Debug.log("GET " + uri);
    HttpRequest req = HttpRequest.newBuilder(URI.create(uri))
        .timeout(Duration.ofSeconds(10))
        .header("Accept", "application/json")
        .GET()
        .build();

    String body = send(req).body();
    JsonArray arr = JsonParser.parseString(body).getAsJsonArray();

    List<Movimiento> list = new ArrayList<>();
    for (JsonElement el : arr) {
      JsonObject o = el.getAsJsonObject();
      Movimiento m = new Movimiento();
      m.setCuenta(optStr(o, "cuenta", "cuentaCodigo"));
      m.setNromov(optInt(o, "nromov", "numero"));
      m.setFecha(optStr(o, "fechaAsString", "fecha"));
      m.setTipo(optStr(o, "tipo", "tipoCodigo"));
      m.setAccion(optStr(o, "accion", "accionTexto"));
      m.setImporte(optDouble(o, "importe"));
      list.add(m);
    }
    Debug.log("Movimientos recibidos: " + list.size());
    return list;
  }

  @Override
  public OperacionCuentaResponse regDeposito(String cuenta, double importe) throws Exception {
    return doOperacion("/deposito", Map.of(
        "cuenta", cuenta,
        "importe", String.valueOf(importe)));
  }

  @Override
  public OperacionCuentaResponse regRetiro(String cuenta, double importe) throws Exception {
    return doOperacion("/retiro", Map.of(
        "cuenta", cuenta,
        "importe", String.valueOf(importe)));
  }

  @Override
  public OperacionCuentaResponse regTransferencia(String origen, String destino, double importe) throws Exception {
    return doOperacion("/transferencia", Map.of(
        "cuentaOrigen", origen,
        "cuentaDestino", destino,
        "importe", String.valueOf(importe)));
  }

  private OperacionCuentaResponse doOperacion(String path, Map<String, String> params) throws Exception {
    String uri = AppConfig.BASE_URL + path + "?" + qs(params);
    Debug.log("POST " + uri);
    HttpRequest req = HttpRequest.newBuilder(URI.create(uri))
        .timeout(Duration.ofSeconds(10))
        .header("Accept", "application/json")
        .POST(HttpRequest.BodyPublishers.noBody())
        .build();

    String body = send(req).body();
    JsonObject o = JsonParser.parseString(body).getAsJsonObject();
    OperacionCuentaResponse r = new OperacionCuentaResponse();
    r.setEstado(o.has("estado") ? o.get("estado").getAsInt() : -1);
    r.setSaldo(o.has("saldo") ? o.get("saldo").getAsDouble() : -1);
    Debug.log("Respuesta op estado=" + r.getEstado() + " saldo=" + r.getSaldo());
    return r;
  }

  private static String optStr(JsonObject o, String... keys) {
    for (String k : keys) if (o.has(k) && !o.get(k).isJsonNull()) return o.get(k).getAsString();
    return "";
  }

  private static int optInt(JsonObject o, String... keys) {
    for (String k : keys) if (o.has(k) && !o.get(k).isJsonNull()) return o.get(k).getAsInt();
    return 0;
  }

  private static double optDouble(JsonObject o, String... keys) {
    for (String k : keys) if (o.has(k) && !o.get(k).isJsonNull()) return o.get(k).getAsDouble();
    return 0;
  }
}
